File: frontend/src/services/knowledge.ts
import { request } from './api'

export type KnowledgeResourceRow = {
  id: string
  chapter_id: string
  chapter_title: string
  title: string
  resource_type: string
  total_chunk_chars: number | string
  chunk_count: number | string
  embedded_count: number | string
}

export type ListKnowledgeResourcesResp = {
  page: number
  page_size: number
  total: number | string
  items: KnowledgeResourceRow[]
}

export type ChunkPreviewSegment = {
  index: number
  content: string
  char_start: number
  char_end: number
}

export type ChunkPreviewResp = {
  segments: ChunkPreviewSegment[]
}

export type ChunkSaveItem = {
  content: string
  char_start?: number
  char_end?: number
}

export async function listKnowledgeResources(token: string, courseId: string, page: number, pageSize: number) {
  const q = new URLSearchParams({ page: String(page), page_size: String(pageSize) })
  return request<ListKnowledgeResourcesResp>(`/teacher/courses/${courseId}/knowledge/resources?${q}`, {
    method: 'GET',
    token,
  })
}

export async function chunkPreview(
  token: string,
  resourceId: string,
  chunkSize: number,
  overlap: number,
  clearPersistedFirst: boolean,
) {
  return request<ChunkPreviewResp>(`/teacher/resources/${resourceId}/knowledge/chunk-preview`, {
    method: 'POST',
    token,
    body: { chunk_size: chunkSize, overlap, clear_persisted_first: clearPersistedFirst },
  })
}

export async function saveKnowledgeChunks(token: string, resourceId: string, chunks: ChunkSaveItem[]) {
  return request<unknown>(`/teacher/resources/${resourceId}/knowledge/chunks`, {
    method: 'PUT',
    token,
    body: { chunks },
  })
}

export async function confirmKnowledgeChunks(token: string, resourceId: string) {
  return request<{ confirmed_count: number }>(`/teacher/resources/${resourceId}/knowledge/chunks/confirm`, {
    method: 'POST',
    token,
  })
}

export type SavedChunkRow = {
  id: string
  chunk_index: number
  content: string
  char_start?: number | null
  char_end?: number | null
  confirmed_at?: string | null
  embedded_at?: string | null
  embedding_dims?: number | null
}

export async function listKnowledgeChunks(token: string, resourceId: string) {
  return request<{ items: SavedChunkRow[] }>(`/teacher/resources/${resourceId}/knowledge/chunks`, {
    method: 'GET',
    token,
  })
}

export async function clearKnowledgeChunks(token: string, resourceId: string) {
  return request<unknown>(`/teacher/resources/${resourceId}/knowledge/chunks`, {
    method: 'DELETE',
    token,
  })
}

export async function updateKnowledgeChunk(
  token: string,
  resourceId: string,
  chunkId: string,
  body: { content: string; char_start?: number; char_end?: number },
) {
  return request<unknown>(`/teacher/resources/${resourceId}/knowledge/chunks/${chunkId}`, {
    method: 'PATCH',
    token,
    body,
  })
}

export async function deleteKnowledgeChunk(token: string, resourceId: string, chunkId: string) {
  return request<unknown>(`/teacher/resources/${resourceId}/knowledge/chunks/${chunkId}`, {
    method: 'DELETE',
    token,
  })
}

export async function embedResource(token: string, resourceId: string, modelId: string) {
  return request<{ embedded_count: number }>(`/teacher/resources/${resourceId}/knowledge/embed`, {
    method: 'POST',
    token,
    body: { model_id: Number(modelId) },
  })
}

File: internal/dto/knowledge/knowledge.go
// Package knowledge 存放课程知识库分块与嵌入相关 DTO。
package knowledge

// ListKnowledgeResourcesResp 课程下可解析资源分页列表（含分块统计）。
type ListKnowledgeResourcesResp struct {
	Page     int              `json:"page"`
	PageSize int              `json:"page_size"`
	Total    int64            `json:"total"`
	Items    []map[string]any `json:"items"`
}

// ChunkPreviewReq 分块预览参数（基于 docreader 解析后的 Markdown）。
type ChunkPreviewReq struct {
	ChunkSize int `json:"chunk_size" binding:"required,min=1,max=32000"`
	Overlap   int `json:"overlap" binding:"min=0"`
	// ClearPersistedFirst 为 true 时先删除该资源下已落库的分块与向量，再解析预览，避免与历史持久化结果混用。
	ClearPersistedFirst bool `json:"clear_persisted_first"`
}

// UpdateKnowledgeChunkReq 更新已保存分块正文；会清空该条向量与确认状态以便重新确认/嵌入。
type UpdateKnowledgeChunkReq struct {
	Content   string `json:"content" binding:"required"`
	CharStart *int   `json:"char_start,omitempty"`
	CharEnd   *int   `json:"char_end,omitempty"`
}

// ChunkPreviewSegment 预览中的单个分块。
type ChunkPreviewSegment struct {
	Index     int    `json:"index"`
	Content   string `json:"content"`
	CharStart int    `json:"char_start"`
	CharEnd   int    `json:"char_end"`
}

// ChunkPreviewResp 分块预览结果。
type ChunkPreviewResp struct {
	Segments []ChunkPreviewSegment `json:"segments"`
}

// ChunkSaveItem 落库分块一行；chunk_index 由服务端按数组顺序从 0 递增写入。
type ChunkSaveItem struct {
	Content   string         `json:"content" binding:"required"`
	CharStart *int           `json:"char_start,omitempty"`
	CharEnd   *int           `json:"char_end,omitempty"`
	Metadata  map[string]any `json:"metadata,omitempty"`
}

// SaveKnowledgeChunksReq 覆盖保存某资源下全部分块（会删除旧分块与向量）；chunks 为空表示清空该资源全部分块。
type SaveKnowledgeChunksReq struct {
	Chunks []ChunkSaveItem `json:"chunks"`
}

// EmbedResourceReq 对已确认且未嵌入的分块调用外部嵌入 API。
type EmbedResourceReq struct {
	ModelID int64 `json:"model_id" binding:"required"`
}

// EmbedResourceResp 嵌入任务摘要。
type EmbedResourceResp struct {
	EmbeddedCount int `json:"embedded_count"`
}

// ConfirmKnowledgeChunksResp 确认分块后的行数。
type ConfirmKnowledgeChunksResp struct {
	ConfirmedCount int64 `json:"confirmed_count"`
}

// ListKnowledgeChunksResp 已保存分块列表。
type ListKnowledgeChunksResp struct {
	Items []map[string]any `json:"items"`
}
